staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1180 @@
+	// informer is not associated to a single VNI. Pick 0 because it's not a
+	// valid VNI value: there's no overlapping with the other IDs.
+	localAttsCacheID cacheID = 0
+
+	// Name of the indexer used to match pre-existing network interfaces to
+	// network attachments.
+	ifcOwnerDataIndexerName = "ifcOwnerData"
+
+	// Field names of NetworkAttachments used to build field selectors.
+	attNodeField   = "spec.node"
+	attIPv4Field   = "status.ipv4"
+	attHostIPField = "status.hostIP"
+	attVNIField    = "status.addressVNI"
+
+	// resync period for Informers caches. Set
+	// to 0 because we don't want resyncs.
+	resyncPeriod = 0
+
+	// netFabricRetryPeriod is the time we wait before retrying when a
+	// network fabric operation fails while handling pre-existing interfaces.
+	netFabricRetryPeriod = time.Second
+
+	// The HTTP port at which Prometheus metrics are served.
+	// Pick an unusual one because the host's network namespace is used.
+	// See https://github.com/prometheus/prometheus/wiki/Default-port-allocations .
+	metricsAddr = ":9294"
+
+	// The HTTP path at which Prometheus metrics are served.
+	metricsPath = "/metrics"
+
+	// The namespace and subsystem of the Prometheus metrics produced here
+	metricsNamespace = "kos"
+	metricsSubsystem = "agent"
+)
+
+// cacheID models the ID of an informer's cache. Used to keep track of which
+// informer's cache NetworkAttachments were seen in and should be looked up in.
+type cacheID uint32
+
+// layer1VirtualNetworkState is the first layer of state associated with a
+// single relevant virtual network.
+type layer1VirtualNetworkState struct {
+	// unique identifier over a run of the connection agent of this
+	// layer1VirtualNetworkState. Used to prevent race conditions.
+	uid uint64
+
+	// List of names of remote NetworkAttachments in the virtual network (which
+	// implicitly determines the namespace) for whom an add notification handler
+	// has been executed but a delete notification handler has not.
+	remoteAtts map[string]struct{}
+
+	// Lister used by workers to retrieve the NetworkAttachment they're
+	// processing.
+	remoteAttsLister koslisterv1a1.NetworkAttachmentNamespaceLister
+}
+
+// layer1VirtualNetworksState is the first layer of state associated with all
+// the relevant virtual networks.
+// Its main purposes are retrieval of remote NetworkAttachments by workers and
+// deletion of network interfaces of those remote NetworkAttachments when the
+// virtual network becomes irrelevant.
+// All operations on a layer1VirtualNetworksState must be done while holding
+// its mutex's lock.
+type layer1VirtualNetworksState struct {
+	sync.RWMutex
+
+	// attToSeenInCaches maps NetworkAttachments namespaced names to the list
+	// of IDs of the Informer's caches where the attachments have been seen.
+	attToSeenInCaches map[k8stypes.NamespacedName]map[cacheID]struct{}
+
+	// vniToVNState maps a VNI to its layer1VirtualNetworkState.
+	vniToVNState map[uint32]*layer1VirtualNetworkState
```

Comment:
> These two maps reveal essentially the same relation from two different directions.

Yes, each direction is meant to simplify code for different scenarios. `attToSeenInCaches` allows fast evaluation of the informer's cache where a networkattachment should be looked up. `vniToVNState` makes it straightforward to enqueue the namespaced names of the remote NetworkAttachments when a virtual network becomes irrelevant. Having only one "direction" would make the code for the non-covered scenario longer (while holding a mutex).

> Note that one field speaks of "cacheID" while the other speaks of "VNI".

That's because NetworkAttachments can be seen in Informer's caches, and there's not a 1 to 1 mapping between Informers' caches and VNIs because of the informer for local attachments. Each virtual network state OTOH is associated to a VNI, not to a cache ID.